src/solvers/smt2/smt2_solver.cpp

```diff
@@ -81,39 +81,42 @@ void smt2_solvert::expand_function_applications(exprt &expr)
   {
     auto &app=to_function_application_expr(expr);
 
-    // look it up
-    irep_idt identifier=app.function().get_identifier();
-    auto f_it=id_map.find(identifier);
-
-    if(f_it!=id_map.end())
+    if(app.function().id() == ID_symbol)
     {
-      const auto &f=f_it->second;
+      // look up the symbol
+      auto identifier = to_symbol_expr(app.function()).get_identifier();
+      auto f_it = id_map.find(identifier);
 
-      DATA_INVARIANT(f.type.id()==ID_mathematical_function,
-        "type of function symbol must be mathematical_function_type");
+      if(f_it != id_map.end())
+      {
+        const auto &f = f_it->second;
 
-      const auto f_type=
-        to_mathematical_function_type(f.type);
+        DATA_INVARIANT(
+          f.type.id() == ID_mathematical_function,
+          "type of function symbol must be mathematical_function_type");
 
-      const auto &domain = f_type.domain();
+        const auto f_type = to_mathematical_function_type(f.type);
 
-      DATA_INVARIANT(
-        domain.size() == app.arguments().size(),
-        "number of function parameters");
+        const auto &domain = f_type.domain();
 
-      replace_symbolt replace_symbol;
+        DATA_INVARIANT(
+          domain.size() == app.arguments().size(),
+          "number of function parameters");
 
-      std::map<irep_idt, exprt> parameter_map;
-      for(std::size_t i = 0; i < domain.size(); i++)
-      {
-        const symbol_exprt s(f.parameters[i], domain[i]);
-        replace_symbol.insert(s, app.arguments()[i]);
-      }
+        replace_symbolt replace_symbol;
 
-      exprt body=f.definition;
-      replace_symbol(body);
-      expand_function_applications(body);
-      expr=body;
+        std::map<irep_idt, exprt> parameter_map;
+        for(std::size_t i = 0; i < domain.size(); i++)
+        {
+          const symbol_exprt s(f.parameters[i], domain[i]);
+          replace_symbol.insert(s, app.arguments()[i]);
+        }
+
+        exprt body = f.definition;
+        replace_symbol(body);
+        expand_function_applications(body);
+        expr = body;
```

Comment:
Why declare an intermediary expression body instead of assigning first to expr and doing the operations directly on it?